dosbox-staging/dosbox-staging
# Fix improper handling of some characters in shell history

# Description

DRAFT for now, because I'm still testing this code.

I have discovered a small shell history related bug:
- STEP: Press <LEFT CONTROL>+a, <ENTER>
- RESULT: A smiley face is printed and illegal command error (OK)
- RESULT: A log is printed, _UNICODE: No fallback mapping for code point 0x263a_ (NOT OK)
- STEP: Press <CURSOR UP> to bring back the last command
- RESULT: A question mark is brought back instead of the smiley face (not OK)

![Screenshot](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/25d99c4e-fc00-40ff-b581-4fcb0c849553)

Solution: the Unicode engine needs to be precisely told what kind of DOS string do we have - with control codes, with screen codes, or with no special characters like that at all.
The Unicode extension implemented here might be helpful for the future use cases too - for UTF8 host file names, or for the upcoming (hopefully someday) EDIT.COM replacement.


# Manual testing

- try to reproduce the problem according to steps within the PR description; make sure that `shell_history_file` config option is not empty check that after restarting DOSBox the shell history is preserved, including the smiley
- check that `tree` command works without graphic glitches
- execute `keyb br 3848` command and check that the `tree` command still works, this time with fallback artwork
- check that `intro` command still works and does not produce graphical glitches
- press `CTRL+F6`, pay attention to the animated recording mark on the title bar and task bar (if your OS has one...) - there should be no regression here
- set `language = pl` config option, there should be no graphical glitches in the default tittle bar (but a couple of Polish national characters)
- set `language = pl` config option, type execute `more z:\autoexec.bat` - there should be no graphical glitches
- with language set to Polish (as above), execute `keyb pl` and `more z:\autoexec.bat` - there should be no graphical glitches, the first line of the file shall end with _ń_ letter (accented n)
- execute 'config -wcd' command with non-empty `[autoexec]` section; config file should be written correctly, including the `[autoexec]` section
- execute `config -axadd FOO`, and `config -axtype` afterwards - content of the updated `[autoexec]` section should be correctly printed out
- display help of any command, for example execute `dir /?`


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:
